real_time/add-custom-dynamic-entity-data-source/src/main/java/com/esri/samples/add_custom_dynamic_entity_data_source/AddCustomDynamicEntityDataSourceSample.java

```diff
@@ -0,0 +1,151 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.esri.samples.add_custom_dynamic_entity_data_source;
+
+import com.esri.arcgisruntime.ArcGISRuntimeEnvironment;
+import com.esri.arcgisruntime.arcgisservices.LabelDefinition;
+import com.esri.arcgisruntime.geometry.Point;
+import com.esri.arcgisruntime.geometry.SpatialReferences;
+import com.esri.arcgisruntime.layers.DynamicEntityLayer;
+import com.esri.arcgisruntime.mapping.ArcGISMap;
+import com.esri.arcgisruntime.mapping.BasemapStyle;
+import com.esri.arcgisruntime.mapping.Viewpoint;
+import com.esri.arcgisruntime.mapping.labeling.SimpleLabelExpression;
+import com.esri.arcgisruntime.mapping.view.MapView;
+import com.esri.arcgisruntime.symbology.TextSymbol;
+
+import javafx.application.Application;
+import javafx.scene.Scene;
+import javafx.scene.layout.StackPane;
+import javafx.scene.paint.Color;
+import javafx.stage.Stage;
+
+public class AddCustomDynamicEntityDataSourceSample extends Application {
+
+  private MapView mapView = new MapView();
+
+  private SimulatedDataSource dynamicEntityDataSource;
+
+  @Override
+  public void start(Stage stage) {
+    StackPane stackPane = new StackPane();
+    Scene scene = new Scene(stackPane);
+
+    // Set title, size, and add scene to stage
+    stage.setTitle("Add custom dynamic entity data source");
+    stage.setWidth(800);
+    stage.setHeight(700);
+    stage.setScene(scene);
+    stage.show();
+
+    // Authentication with an API key or named user is required to access basemaps and other location services
+    String yourAPIKey = System.getProperty("apiKey");
+    ArcGISRuntimeEnvironment.setApiKey(yourAPIKey);
+
+    // Create a new map with the navigation basemap style.
+    ArcGISMap map = new ArcGISMap(BasemapStyle.ARCGIS_OCEANS);
+
+    // Create a map view and set the map to it
+    mapView = new MapView();
+    mapView.setMap(map);
+
+    // Set the initial viewpoint.
+    mapView.setViewpoint(new Viewpoint(new Point(-123.657, 47.984, SpatialReferences.getWgs84()), 3e6));
+
+    // Create a custom data source implementation of a DynamicEntityDataSource.
+    // This takes the path to the simulation file, field name that will be used as the entity id, and the delay between
+    // each observation that is processed.
+    // In this example we are using a json file (stored as a resource) as our custom data source.
+    // This field value should be a unique identifier for each entity.
+    // Adjusting the value for the delay will change the speed at which the entities and their observations are displayed.
+    var resource = getClass().getResource("/add_custom_dynamic_entity_data_source/AIS_MarineCadastre_SelectedVessels_CustomDataSource.json");
+    dynamicEntityDataSource = new SimulatedDataSource(resource.getPath(), "MMSI", 10);
```

Comment:
This is quite a lot of comment - feels like we're repeating the Javadoc for `SimulatedDataSource`'s constructor a bit. Think it could do with being reduced and split up a bit, something like:
```suggestion
    // JSON file (stored as a resource) for use as custom data source
    var resource = getClass().getResource("/add_custom_dynamic_entity_data_source/AIS_MarineCadastre_SelectedVessels_CustomDataSource.json");
    // create a custom data source implementation of a DynamicEntityDataSource with a data source, entity id field name, and update delay
    dynamicEntityDataSource = new SimulatedDataSource(resource.getPath(), "MMSI", 10);
```

I'm not saying to write this specifically but see what you can do.

Reply:
Also wondering if it's better to have the call to `getPath` on the line assigning `resource`. It's a bit of an ambiguous  `var` there, not one I think we'd have in other places (I could be wrong!)

Reply:
Updated it to be more like what you suggest.